server/src/com/thoughtworks/go/server/service/EnvironmentConfigService.java

```diff
@@ -145,6 +136,24 @@ public EnvironmentConfig named(String environmentName) throws NoSuchEnvironmentE
         return environments.named(new CaseInsensitiveString(environmentName));
     }
 
+    public EnvironmentConfig getEnvironmentConfig(String environmentName) throws NoSuchEnvironmentException {
+        if(environments.hasEnvironmentNamed(new CaseInsensitiveString(environmentName)))
+            return environments.named(new CaseInsensitiveString(environmentName));
+        return null;
+    }
+
+    public void deleteEnvironment(String environmentName, HttpLocalizedOperationResult result) throws NoSuchEnvironmentException {
+        CruiseConfig config = goConfigService.getConfigForEditing();
+        EnvironmentsConfig environments = config.getEnvironments();
+        if (environments.remove(getEnvironmentConfig(environmentName))) {
+            config.setEnvironments(environments);
```

Comment:
Race condition here —

If another thread were to do add/remove an environment after you store the original environments using `config.getEnvironments` this call to `config.setEnvironments()` will get rid of those environments that were deleted, or add back environments that were removed (essentially rolling back any changes by the other thread)